core/storage/question/gae_models.py

```diff
@@ -0,0 +1,114 @@
+    """Storage model for the metadata for a question snapshot."""
+    pass
+
+
+class QuestionSnapshotContentModel(base_models.BaseSnapshotContentModel):
+    """Storage model for the content of a question snapshot."""
+    pass
+
+
+class QuestionModel(base_models.VersionedModel):
+    """Model for storing Questions.
+
+    The id of instances of this class has the form
+    {{collection_id}}.{{random_hash_of_16_chars}}
+    """
+    SNAPSHOT_METADATA_CLASS = QuestionSnapshotMetadataModel
+    SNAPSHOT_CONTENT_CLASS = QuestionSnapshotContentModel
+    ALLOW_REVERT = True
+
+    # The title of the question.
+    title = ndb.StringProperty(required=True, indexed=True)
+    # The data for the question data.
+    question_data = ndb.JsonProperty(default={}, indexed=False)
+    # The schema version for the data.
+    data_schema_version = ndb.IntegerProperty(required=True, indexed=True)
+    # The ID of collection to which the question belongs.
+    collection_id = ndb.StringProperty(required=True, indexed=True)
+    # The ISO 639-1 code for the language this question is written in.
+    language_code = ndb.StringProperty(required=True, indexed=True)
+
+    @classmethod
+    def _get_new_id(cls, collection_id):
+        """Generates a unique id for the question of the form
+        {{collection_id}}.{{random_hash_of_16_chars}}
+
+        Args:
+            collection_id: str. ID of the collection.
+
+        Returns:
+           new_id: int. ID of the new QuestionModel instance.
+
+        Raises:
+            Exception: The id generator for QuestionModel is
+            producing too many collisions.
+        """
+
+        for _ in range(base_models.MAX_RETRIES):
+            new_id = '%s.%s' % (
+                collection_id,
+                utils.convert_to_hash(
+                    str(utils.get_random_int(base_models.RAND_RANGE)),
+                    base_models.ID_LENGTH))
+            if not cls.get_by_id(new_id):
+                return new_id
+
+        raise Exception(
+            'The id generator for QuestionModel is producing too many '
+            'collisions.')
+
+    @classmethod
+    def create(
+            cls, title, question_data, data_schema_version, collection_id,
+            language_code):
+        """Creates a new QuestionModel entry.
+
+        Args:
+            title: str. The title of the question.
+            question_data: dict. The data for the question data.
+            data_schema_version: int. The schema version for the data.
+            collection_id: str. ID of the collection.
+            language_code: str. The ISO 639-1 code for the language this
+                question is written in.
+
+        Returns:
+            question_instance: QuestionModel. Instance of the new QuestionModel
+                entry.
+
+        Raises:
+            Exception: A model with the same ID already exists.
+        """
+        instance_id = cls._get_new_id(collection_id)
+        question_instance = cls(
```

Comment:
Let's call this question_model_instance to distinguish from the domain object.

Reply:
Updated.